### ISSUE TITLE
ffigen should not generate if input mtimes are older than output mtimes

### ISSUE THREAD

**Issue by johnmccutchan** [dart-lang/native#879]
ffigen should early out if the input header/config mtimes are older than the generated code mtimes

**Comment by PranjalManhgaye** [comment]
Hiiii  i'd like to work on this issue. I plan to add a timestamp check to skip generation when outputs are newer than inputs. Please let me know if that sounds right , can i be assigned with this issue

**Comment by dcharkes** [comment]
@PranjalManhgaye https://github.com/dart-lang/native/issues/1008 should be addressed first.

**Comment by liamappelbe** [comment]
I don't think this can be done accurately for the Dart API case, so it should either be yaml config only, or (imo) not done at all since we're planning to deprecate the yaml config.

**Comment by dcharkes** [comment]
> I don't think this can be done accurately for the Dart API case

Hm, right we'd need to also read the Dart source itself. (We could potentially deep-hash all collections/strings/etc. But we cannot hash callbacks and visitors.)

We could read the Dart sources and add some logic to the generate hooks. Then the generate hook would be responsible for determining whether the Dart sources change, and the hook itself would be responsible for tracking its own inputs. If we also add a list of dependencies as output like we do for build hooks, we can make `dart generate` skip the generate hook if neither the C/Objc-C sources changed nor the Dart sources. Or an alternative design would be we always invoke the hook but the `GenerateInput` has a `bool dartSourcesChanged`, and then the internal (optional) caching inside the hook can use that information.

I agree that fixing this is blocked until we have generate hooks (https://github.com/dart-lang/sdk/issues/56512).

We can still work on https://github.com/dart-lang/native/issues/1008, outputting all files that clang read.
